File: engine/src/rules/evaluators.ts
/**
 * Deterministic evaluators. Each one maps to a section of the delivery standard
 * and is registered under a stable key that a {@link ReviewRule.evaluator}
 * references. Every threshold, name and list is read from the rule's
 * `parameters`, so behaviour is configured per tenant, not hard-coded.
 */
import type { Evaluator, EvaluatorContext, FlowAction, Finding } from '../types.js';
import {
  actionParameters,
  apiId,
  arrParam,
  asRecord,
  boolParam,
  escapeRegex,
  finding,
  operationId,
  sameStatusSet,
  strParam,
} from './helpers.js';

/** Every top-level Scope action, keyed case-insensitively by name. */
function topLevelScopes(ctx: EvaluatorContext): Map<string, FlowAction> {
  const map = new Map<string, FlowAction>();
  for (const a of ctx.flow.topLevelActions) {
    if (a.type.toLowerCase() === 'scope') map.set(a.name.toLowerCase(), a);
  }
  return map;
}

/**
 * NAMING_CONVENTION — flow name matches `<Prefix><sep><Area><sep><Function>`.
 * Params: `prefix`, `separator`, `minSegments`, or an explicit `pattern`.
 */
export const namingConvention: Evaluator = (ctx) => {
  const p = ctx.rule.parameters;
  const explicit = strParam(p, 'pattern', '');
  const prefix = strParam(p, 'prefix', 'CMA');
  const separator = strParam(p, 'separator', ' - ');
  const minSegments = typeof p.minSegments === 'number' ? p.minSegments : 3;

  let regex: RegExp;
  if (explicit) {
    regex = new RegExp(explicit);
  } else {
    const tail = Array.from({ length: Math.max(0, minSegments - 1) }, () => '.+').join(
      escapeRegex(separator),
    );
    regex = new RegExp(`^${escapeRegex(prefix)}${escapeRegex(separator)}${tail}$`);
  }

  const name = ctx.flow.displayName;
  if (regex.test(name)) {
    return [finding(ctx, 'pass', `Name "${name}" follows the naming convention.`)];
  }
  return [
    finding(
      ctx,
      'fail',
      `Name "${name}" does not match the required pattern ${regex.source}. Expected e.g. "${prefix}${separator}<Area>${separator}<Function>".`,
      { evidence: name },
    ),
  ];
};

/**
 * ERROR_HANDLING_SCOPES — Try / Catch / Finally scope pattern with the correct
 * `runAfter` wiring. This is the highest-signal structural check.
 * Params: `tryName`, `catchName`, `finallyName`, `requireFinally`,
 * `catchRunAfterStatuses`, `finallyRunAfterStatuses`.
 */
export const errorHandlingScopes: Evaluator = (ctx) => {
  const p = ctx.rule.parameters;
  const tryName = strParam(p, 'tryName', 'Try');
  const catchName = strParam(p, 'catchName', 'Catch');
  const finallyName = strParam(p, 'finallyName', 'Finally');
  const requireFinally = boolParam(p, 'requireFinally', true);
  const catchStatuses = arrParam(p, 'catchRunAfterStatuses', ['Failed', 'TimedOut', 'Skipped']);
  const finallyStatuses = arrParam(p, 'finallyRunAfterStatuses', [
    'Succeeded',
    'Failed',
    'TimedOut',
    'Skipped',
  ]);

  const scopes = topLevelScopes(ctx);
  const tryScope = scopes.get(tryName.toLowerCase());
  const catchScope = scopes.get(catchName.toLowerCase());
  const finallyScope = scopes.get(finallyName.toLowerCase());

  const problems: string[] = [];

  if (!tryScope) problems.push(`missing a top-level "${tryName}" Scope`);
  if (!catchScope) {
    problems.push(`missing a top-level "${catchName}" Scope`);
  } else {
    const ra = catchScope.runAfter[tryName] ?? catchScope.runAfter[tryScope?.name ?? tryName];
    if (!ra) {
      problems.push(`"${catchName}" does not run after "${tryName}"`);
    } else if (!sameStatusSet(ra, catchStatuses)) {
      problems.push(
        `"${catchName}" runAfter is [${ra.join(', ')}]; expected [${catchStatuses.join(', ')}]`,
      );
    }
  }
  if (requireFinally) {
    if (!finallyScope) {
      problems.push(`missing a top-level "${finallyName}" Scope`);
    } else {
      const ra = finallyScope.runAfter[catchName];
      if (!ra) {
        problems.push(`"${finallyName}" does not run after "${catchName}"`);
      } else if (!sameStatusSet(ra, finallyStatuses)) {
        problems.push(
          `"${finallyName}" runAfter is [${ra.join(', ')}]; expected [${finallyStatuses.join(', ')}]`,
        );
      }
    }
  }

  if (problems.length === 0) {
    return [finding(ctx, 'pass', 'Try / Catch / Finally scopes are present and correctly wired.')];
  }
  return [
    finding(ctx, 'fail', `Error-handling pattern is incomplete: ${problems.join('; ')}.`, {
      evidence: [...scopes.values()].map((s) => s.name).join(', ') || '(no scopes)',
    }),
  ];
};

/**
 * LOGGING_PRESENT — the flow calls the shared logging child flow.
 * Params: `loggerNameContains` (substrings), `loggerWorkflowIds`.
 */
export const loggingPresent: Evaluator = (ctx) => {
  const p = ctx.rule.parameters;
  const names = arrParam(p, 'loggerNameContains', ['Logger', 'Log']).map((s) => s.toLowerCase());
  const ids = arrParam(p, 'loggerWorkflowIds', []).map((s) => s.toLowerCase());

  const match = ctx.flow.allActions.find((a) => {
    const isChildFlow = a.type.toLowerCase() === 'workflow';
    const nameHit = names.some((n) => a.name.toLowerCase().includes(n));
    const host = asRecord(asRecord(a.inputs).host);
    const ref = String(host.workflowReferenceName ?? '').toLowerCase();
    const idHit = ids.length > 0 && ids.includes(ref);
    return (isChildFlow && (nameHit || idHit || ids.length === 0)) || (nameHit && isChildFlow);
  });

  if (match) {
    return [finding(ctx, 'pass', `Logging child flow invoked via "${match.name}".`, {
      evidence: match.path,
    })];
  }
  return [
    finding(
      ctx,
      'fail',
      'No call to the shared logging child flow was found. Long-term history is lost after 28 days.',
    ),
  ];
};

/**
 * CONFIG_LIST_TOP — every Dataverse "List rows" action bounds its query with
 * `$top` (the "Row Count = 1" guidance) to avoid open-ended query warnings.
 * Params: `listOperationIds`, `topParamNames`.
 */
export const configListTop: Evaluator = (ctx) => {
  const p = ctx.rule.parameters;
  const ops = arrParam(p, 'listOperationIds', ['ListRecords']).map((s) => s.toLowerCase());
  const topKeys = arrParam(p, 'topParamNames', ['$top', 'top']);

  const listActions = ctx.flow.allActions.filter((a) => ops.includes(operationId(a).toLowerCase()));
  if (listActions.length === 0) {
    return [finding(ctx, 'not_applicable', 'No Dataverse list actions in this flow.')];
  }

  const offenders = listActions.filter((a) => {
    const params = actionParameters(a);
    return !topKeys.some((k) => params[k] !== undefined && params[k] !== null && params[k] !== '');
  });

  if (offenders.length === 0) {
    return [
      finding(ctx, 'pass', `All ${listActions.length} list action(s) bound with a row count.`),
    ];
  }
  return offenders.map((a) =>
    finding(
      ctx,
      'fail',
      `List action "${a.name}" has no row-count limit (${topKeys.join('/')}). Open-ended queries raise Microsoft warnings.`,
      { evidence: a.path },
    ),
  );
};

/**
 * APPROVED_EMAIL_SENDER — outbound email uses an approved sender/connection,
 * never a personal account.
 * Params: `emailOperationIds`, `emailApiIdContains`, `approvedSenders`,
 * `fromParamNames`.
 */
export const approvedEmailSender: Evaluator = (ctx) => {
  const p = ctx.rule.parameters;
  const ops = arrParam(p, 'emailOperationIds', [
    'SendEmailV2',
    'SendEmail',
    'SharedMailboxSendEmailV2',
  ]).map((s) => s.toLowerCase());
  const apiContains = arrParam(p, 'emailApiIdContains', ['office365', 'outlook', 'gmail']).map((s) =>
    s.toLowerCase(),
  );
  const approved = arrParam(p, 'approvedSenders', ['appdev-no-reply@cma.ca']).map((s) =>
    s.toLowerCase(),
  );
  const fromKeys = arrParam(p, 'fromParamNames', ['emailMessage/From', 'From', 'MailboxAddress']);

  const emailActions = ctx.flow.allActions.filter((a) => {
    const op = operationId(a).toLowerCase();
    const api = apiId(a).toLowerCase();
    return ops.includes(op) || (apiContains.some((c) => api.includes(c)) && op.includes('email'));
  });

  if (emailActions.length === 0) {
    return [finding(ctx, 'not_applicable', 'This flow does not send email.')];
  }

  const findings: Finding[] = [];
  for (const a of emailActions) {
    const params = actionParameters(a);
    const fromKey = fromKeys.find((k) => typeof params[k] === 'string' && params[k] !== '');
    if (!fromKey) {
      // Sender is the connection identity, not visible in the definition.
      findings.push(
        finding(
          ctx,
          'warning',
          `Email action "${a.name}" does not declare a sender in its definition. Verify the connection uses an approved sender (${approved.join(', ')}).`,
          { evidence: a.path },
        ),
      );
      continue;
    }
    const from = String(params[fromKey]).toLowerCase();
    if (approved.some((s) => from.includes(s))) {
      findings.push(finding(ctx, 'pass', `Email action "${a.name}" uses an approved sender.`, {
        evidence: a.path,
      }));
    } else {
      findings.push(
        finding(
          ctx,
          'fail',
          `Email action "${a.name}" sends from "${params[fromKey]}", which is not an approved sender (${approved.join(', ')}).`,
          { evidence: a.path },
        ),
      );
    }
  }
  return findings;
};

/**
 * RUN_AS_SERVICE_PRINCIPAL — the flow should not run as a user principal.
 * Uses inventory metadata (owner type) enriched from the Management API.
 */
export const runAsServicePrincipal: Evaluator = (ctx) => {
  const owner = ctx.inventory?.ownerType;
  if (!owner || owner === 'unknown') {
    return [finding(ctx, 'not_applicable', 'Owner principal type unknown; cannot evaluate.')];
  }
  if (owner === 'user') {
    return [
      finding(
        ctx,
        'fail',
        `Flow runs as a user principal (${ctx.inventory?.ownerName ?? 'user'}). It should run as an application / service principal.`,
      ),
    ];
  }
  return [finding(ctx, 'pass', `Flow runs as a ${owner} principal.`)];
};

/**
 * NO_HARDCODED_ENV — heuristic scan for environment-specific literals (absolute
 * https URLs, environment GUIDs) that should come from the config table.
 * Params: `ignoreHostSubstrings`, `guidCheck`.
 */
export const noHardcodedEnv: Evaluator = (ctx) => {
  const p = ctx.rule.parameters;
  const ignore = arrParam(p, 'ignoreHostSubstrings', [
    'schema.management.azure.com',
    'schemas.microsoft.com',
    'login.microsoftonline.com',
  ]).map((s) => s.toLowerCase());
  const checkGuid = boolParam(p, 'guidCheck', true);

  const urlRe = /https?:\/\/[^\s"'\\)]+/gi;
  const guidRe = /\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\b/gi;

  const hits: string[] = [];
  for (const a of ctx.flow.allActions) {
    // Only scan the parameters block; connector host URLs are legitimate.
    const params = JSON.stringify(actionParameters(a) ?? {});
    for (const m of params.match(urlRe) ?? []) {
      if (!ignore.some((ig) => m.toLowerCase().includes(ig))) hits.push(`${a.name}: ${m}`);
    }
    if (checkGuid) {
      for (const m of params.match(guidRe) ?? []) hits.push(`${a.name}: ${m}`);
    }
  }

  if (hits.length === 0) {
    return [finding(ctx, 'pass', 'No hard-coded environment-specific literals detected.')];
  }
  return [
    finding(
      ctx,
      'warning',
      `Possible hard-coded environment-specific value(s) found; consider loading from the Flow Configuration table: ${hits
        .slice(0, 5)
        .join('; ')}${hits.length > 5 ? ` (+${hits.length - 5} more)` : ''}.`,
      { evidence: hits[0] },
    ),
  ];
};

/** Registry consumed by the reviewer. Keys match `ReviewRule.evaluator`. */
export const evaluatorRegistry: Record<string, Evaluator> = {
  namingConvention,
  errorHandlingScopes,
  loggingPresent,
  configListTop,
  approvedEmailSender,
  runAsServicePrincipal,
  noHardcodedEnv,
};

File: plugin/Ember.Plugins/Engine/Evaluators.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using BetaRover.Ember.Json;

namespace BetaRover.Ember.Engine
{
    /// <summary>Port of engine/src/rules/helpers.ts + evaluators.ts.</summary>
    public static class Evaluators
    {
        // ---------------- helpers ----------------
        private static Dictionary<string, object> AsRecord(object v) => JsonUtil.AsObject(v);

        private static string StrParam(Dictionary<string, object> p, string key, string fallback)
        {
            object v; return (p.TryGetValue(key, out v) && v is string s) ? s : fallback;
        }

        private static List<string> ArrParam(Dictionary<string, object> p, string key, string[] fallback)
        {
            object v;
            if (p.TryGetValue(key, out v) && v is List<object> arr) return arr.Select(x => JsonUtil.Str(x)).ToList();
            return fallback.ToList();
        }

        private static bool BoolParam(Dictionary<string, object> p, string key, bool fallback)
        {
            object v; return (p.TryGetValue(key, out v) && v is bool b) ? b : fallback;
        }

        private static Dictionary<string, object> ActionHost(FlowAction a) => AsRecord(JsonUtil.Get(AsRecord(a.Inputs), "host"));
        private static Dictionary<string, object> ActionParameters(FlowAction a) => AsRecord(JsonUtil.Get(AsRecord(a.Inputs), "parameters"));
        private static string OperationId(FlowAction a) => JsonUtil.Str(JsonUtil.Get(ActionHost(a), "operationId"));
        private static string ApiId(FlowAction a) => JsonUtil.Str(JsonUtil.Get(ActionHost(a), "apiId"));

        private static bool SameStatusSet(List<string> a, List<string> b)
        {
            var sa = new HashSet<string>(a.Select(s => s.ToLowerInvariant()));
            var sb = new HashSet<string>(b.Select(s => s.ToLowerInvariant()));
            return sa.SetEquals(sb);
        }

        private static string EscapeRegex(string s) => Regex.Replace(s, @"[.*+?^${}()|[\]\\]", "\\$&");

        private static Finding MakeFinding(EvaluatorContext ctx, string status, string message, string evidence = null, string severity = null)
        {
            bool passLike = status == "pass" || status == "not_applicable";
            return new Finding
            {
                RuleCode = ctx.Rule.Code,
                RuleName = ctx.Rule.Name,
                Category = ctx.Rule.Category,
                Status = status,
                Severity = passLike ? "info" : (severity ?? ctx.Rule.Severity),
                Message = message,
                Evidence = evidence,
                Remediation = passLike ? null : ctx.Rule.Remediation,
            };
        }

        private static Dictionary<string, FlowAction> TopLevelScopes(EvaluatorContext ctx)
        {
            var map = new Dictionary<string, FlowAction>();
            foreach (var a in ctx.Flow.TopLevelActions)
                if (a.Type.ToLowerInvariant() == "scope") map[a.Name.ToLowerInvariant()] = a;
            return map;
        }

        // ---------------- evaluators ----------------
        public static List<Finding> NamingConvention(EvaluatorContext ctx)
        {
            var p = ctx.Rule.Parameters;
            string explicitPat = StrParam(p, "pattern", "");
            string prefix = StrParam(p, "prefix", "CMA");
            string separator = StrParam(p, "separator", " - ");
            object ms; int minSegments = (p.TryGetValue("minSegments", out ms) && ms is double d) ? (int)d : 3;

            Regex regex;
            if (!string.IsNullOrEmpty(explicitPat)) regex = new Regex(explicitPat);
            else
            {
                var tail = string.Join(EscapeRegex(separator), Enumerable.Repeat(".+", Math.Max(0, minSegments - 1)));
                regex = new Regex("^" + EscapeRegex(prefix) + EscapeRegex(separator) + tail + "$");
            }

            string name = ctx.Flow.DisplayName ?? "";
            if (regex.IsMatch(name))
                return new List<Finding> { MakeFinding(ctx, "pass", "Name \"" + name + "\" follows the naming convention.") };
            return new List<Finding> { MakeFinding(ctx, "fail",
                "Name \"" + name + "\" does not match the required pattern " + regex.ToString() + ". Expected e.g. \"" + prefix + separator + "<Area>" + separator + "<Function>\".",
                name) };
        }

        public static List<Finding> ErrorHandlingScopes(EvaluatorContext ctx)
        {
            var p = ctx.Rule.Parameters;
            string tryName = StrParam(p, "tryName", "Try");
            string catchName = StrParam(p, "catchName", "Catch");
            string finallyName = StrParam(p, "finallyName", "Finally");
            bool requireFinally = BoolParam(p, "requireFinally", true);
            var catchStatuses = ArrParam(p, "catchRunAfterStatuses", new[] { "Failed", "TimedOut", "Skipped" });
            var finallyStatuses = ArrParam(p, "finallyRunAfterStatuses", new[] { "Succeeded", "Failed", "TimedOut", "Skipped" });

            var scopes = TopLevelScopes(ctx);
            FlowAction tryScope, catchScope, finallyScope;
            scopes.TryGetValue(tryName.ToLowerInvariant(), out tryScope);
            scopes.TryGetValue(catchName.ToLowerInvariant(), out catchScope);
            scopes.TryGetValue(finallyName.ToLowerInvariant(), out finallyScope);

            var problems = new List<string>();
            if (tryScope == null) problems.Add("missing a top-level \"" + tryName + "\" Scope");
            if (catchScope == null) problems.Add("missing a top-level \"" + catchName + "\" Scope");
            else
            {
                List<string> ra;
                if (!catchScope.RunAfter.TryGetValue(tryName, out ra))
                    catchScope.RunAfter.TryGetValue(tryScope != null ? tryScope.Name : tryName, out ra);
                if (ra == null) problems.Add("\"" + catchName + "\" does not run after \"" + tryName + "\"");
                else if (!SameStatusSet(ra, catchStatuses))
                    problems.Add("\"" + catchName + "\" runAfter is [" + string.Join(", ", ra) + "]; expected [" + string.Join(", ", catchStatuses) + "]");
            }
            if (requireFinally)
            {
                if (finallyScope == null) problems.Add("missing a top-level \"" + finallyName + "\" Scope");
                else
                {
                    List<string> ra;
                    finallyScope.RunAfter.TryGetValue(catchName, out ra);
                    if (ra == null) problems.Add("\"" + finallyName + "\" does not run after \"" + catchName + "\"");
                    else if (!SameStatusSet(ra, finallyStatuses))
                        problems.Add("\"" + finallyName + "\" runAfter is [" + string.Join(", ", ra) + "]; expected [" + string.Join(", ", finallyStatuses) + "]");
                }
            }

            if (problems.Count == 0)
                return new List<Finding> { MakeFinding(ctx, "pass", "Try / Catch / Finally scopes are present and correctly wired.") };
            string ev = scopes.Values.Count > 0 ? string.Join(", ", scopes.Values.Select(s => s.Name)) : "(no scopes)";
            return new List<Finding> { MakeFinding(ctx, "fail", "Error-handling pattern is incomplete: " + string.Join("; ", problems) + ".", ev) };
        }

        public static List<Finding> LoggingPresent(EvaluatorContext ctx)
        {
            var p = ctx.Rule.Parameters;
            var names = ArrParam(p, "loggerNameContains", new[] { "Logger", "Log" }).Select(s => s.ToLowerInvariant()).ToList();
            var ids = ArrParam(p, "loggerWorkflowIds", new string[0]).Select(s => s.ToLowerInvariant()).ToList();

            FlowAction match = ctx.Flow.AllActions.FirstOrDefault(a =>
            {
                bool isChildFlow = a.Type.ToLowerInvariant() == "workflow";
                bool nameHit = names.Any(n => a.Name.ToLowerInvariant().Contains(n));
                var host = AsRecord(JsonUtil.Get(AsRecord(a.Inputs), "host"));
                string reference = JsonUtil.Str(JsonUtil.Get(host, "workflowReferenceName")).ToLowerInvariant();
                bool idHit = ids.Count > 0 && ids.Contains(reference);
                return (isChildFlow && (nameHit || idHit || ids.Count == 0)) || (nameHit && isChildFlow);
            });

            if (match != null)
                return new List<Finding> { MakeFinding(ctx, "pass", "Logging child flow invoked via \"" + match.Name + "\".", match.Path) };
            return new List<Finding> { MakeFinding(ctx, "fail",
                "No call to the shared logging child flow was found. Long-term history is lost after 28 days.") };
        }

        public static List<Finding> ConfigListTop(EvaluatorContext ctx)
        {
            var p = ctx.Rule.Parameters;
            var ops = ArrParam(p, "listOperationIds", new[] { "ListRecords" }).Select(s => s.ToLowerInvariant()).ToList();
            var topKeys = ArrParam(p, "topParamNames", new[] { "$top", "top" });

            var listActions = ctx.Flow.AllActions.Where(a => ops.Contains(OperationId(a).ToLowerInvariant())).ToList();
            if (listActions.Count == 0)
                return new List<Finding> { MakeFinding(ctx, "not_applicable", "No Dataverse list actions in this flow.") };

            var offenders = listActions.Where(a =>
            {
                var pr = ActionParameters(a);
                return !topKeys.Any(k => { object val; return pr.TryGetValue(k, out val) && val != null && !"".Equals(val); });
            }).ToList();

            if (offenders.Count == 0)
                return new List<Finding> { MakeFinding(ctx, "pass", "All " + listActions.Count + " list action(s) bound with a row count.") };
            return offenders.Select(a => MakeFinding(ctx, "fail",
                "List action \"" + a.Name + "\" has no row-count limit (" + string.Join("/", topKeys) + "). Open-ended queries raise Microsoft warnings.",
                a.Path)).ToList();
        }

        public static List<Finding> ApprovedEmailSender(EvaluatorContext ctx)
        {
            var p = ctx.Rule.Parameters;
            var ops = ArrParam(p, "emailOperationIds", new[] { "SendEmailV2", "SendEmail", "SharedMailboxSendEmailV2" }).Select(s => s.ToLowerInvariant()).ToList();
            var apiContains = ArrParam(p, "emailApiIdContains", new[] { "office365", "outlook", "gmail" }).Select(s => s.ToLowerInvariant()).ToList();
            var approved = ArrParam(p, "approvedSenders", new[] { "appdev-no-reply@cma.ca" }).Select(s => s.ToLowerInvariant()).ToList();
            var fromKeys = ArrParam(p, "fromParamNames", new[] { "emailMessage/From", "From", "MailboxAddress" });

            var emailActions = ctx.Flow.AllActions.Where(a =>
            {
                string op = OperationId(a).ToLowerInvariant();
                string api = ApiId(a).ToLowerInvariant();
                return ops.Contains(op) || (apiContains.Any(c => api.Contains(c)) && op.Contains("email"));
            }).ToList();

            if (emailActions.Count == 0)
                return new List<Finding> { MakeFinding(ctx, "not_applicable", "This flow does not send email.") };

            var findings = new List<Finding>();
            foreach (var a in emailActions)
            {
                var pr = ActionParameters(a);
                string fromKey = fromKeys.FirstOrDefault(k => { object v; return pr.TryGetValue(k, out v) && v is string sv && sv != ""; });
                if (fromKey == null)
                {
                    findings.Add(MakeFinding(ctx, "warning",
                        "Email action \"" + a.Name + "\" does not declare a sender in its definition. Verify the connection uses an approved sender (" + string.Join(", ", approved) + ").",
                        a.Path));
                    continue;
                }
                string from = JsonUtil.Str(pr[fromKey]).ToLowerInvariant();
                if (approved.Any(s => from.Contains(s)))
                    findings.Add(MakeFinding(ctx, "pass", "Email action \"" + a.Name + "\" uses an approved sender.", a.Path));
                else
                    findings.Add(MakeFinding(ctx, "fail",
                        "Email action \"" + a.Name + "\" sends from \"" + JsonUtil.Str(pr[fromKey]) + "\", which is not an approved sender (" + string.Join(", ", approved) + ").",
                        a.Path));
            }
            return findings;
        }

        public static List<Finding> RunAsServicePrincipal(EvaluatorContext ctx)
        {
            string owner = ctx.Inventory != null ? ctx.Inventory.OwnerType : null;
            if (string.IsNullOrEmpty(owner) || owner == "unknown")
                return new List<Finding> { MakeFinding(ctx, "not_applicable", "Owner principal type unknown; cannot evaluate.") };
            if (owner == "user")
                return new List<Finding> { MakeFinding(ctx, "fail",
                    "Flow runs as a user principal (" + (ctx.Inventory != null && ctx.Inventory.OwnerName != null ? ctx.Inventory.OwnerName : "user") + "). It should run as an application / service principal.") };
            return new List<Finding> { MakeFinding(ctx, "pass", "Flow runs as a " + owner + " principal.") };
        }

        private static readonly Regex UrlRe = new Regex(@"https?://[^\s""'\\)]+", RegexOptions.IgnoreCase);
        private static readonly Regex GuidRe = new Regex(@"\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\b", RegexOptions.IgnoreCase);

        public static List<Finding> NoHardcodedEnv(EvaluatorContext ctx)
        {
            var p = ctx.Rule.Parameters;
            var ignore = ArrParam(p, "ignoreHostSubstrings", new[] { "schema.management.azure.com", "schemas.microsoft.com", "login.microsoftonline.com" }).Select(s => s.ToLowerInvariant()).ToList();
            bool checkGuid = BoolParam(p, "guidCheck", true);

            var hits = new List<string>();
            foreach (var a in ctx.Flow.AllActions)
            {
                string paramsJson = JsonUtil.Serialize(ActionParameters(a));
                foreach (Match m in UrlRe.Matches(paramsJson))
                    if (!ignore.Any(ig => m.Value.ToLowerInvariant().Contains(ig))) hits.Add(a.Name + ": " + m.Value);
                if (checkGuid)
                    foreach (Match m in GuidRe.Matches(paramsJson)) hits.Add(a.Name + ": " + m.Value);
            }

            if (hits.Count == 0)
                return new List<Finding> { MakeFinding(ctx, "pass", "No hard-coded environment-specific literals detected.") };
            string more = hits.Count > 5 ? " (+" + (hits.Count - 5) + " more)" : "";
            return new List<Finding> { MakeFinding(ctx, "warning",
                "Possible hard-coded environment-specific value(s) found; consider loading from the Flow Configuration table: " + string.Join("; ", hits.Take(5)) + more + ".",
                hits[0]) };
        }

        /// <summary>Registry keyed by ReviewRule.Evaluator (matches evaluatorRegistry).</summary>
        public static readonly Dictionary<string, Func<EvaluatorContext, List<Finding>>> Registry =
            new Dictionary<string, Func<EvaluatorContext, List<Finding>>>
            {
                { "namingConvention", NamingConvention },
                { "errorHandlingScopes", ErrorHandlingScopes },
                { "loggingPresent", LoggingPresent },
                { "configListTop", ConfigListTop },
                { "approvedEmailSender", ApprovedEmailSender },
                { "runAsServicePrincipal", RunAsServicePrincipal },
                { "noHardcodedEnv", NoHardcodedEnv },
            };
    }
}
